flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NetworkBufferAllocator.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.io.network.partition.consumer.InputChannelID;
+import org.apache.flink.runtime.io.network.partition.consumer.RemoteInputChannel;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * An allocator used for requesting buffers in the receiver side of netty handlers.
+ */
+public class NetworkBufferAllocator {
+    private final CreditBasedPartitionRequestClientHandler partitionRequestClientHandler;
```

Comment:
It seems tricky to rely on `CreditBasedPartitionRequestClientHandler` in the allocator, because the handler is too large component and which is not used for getting `InputChannel` by design. We can think whether there are better ways to do that.

Reply:
Maybe it is better that we pass `NetworkClientHandler` instead here and add the interface method `getInputChannel` for it if no other good options. Let me think it through later.

Reply:
Agree with that it should be better to rely on the interface instead of implementation. I have modified the code according.

However, One small concern remaining is that the new decoder works only when it ensures the receiver side always have available buffers. The current NetworkClient interface cannot ensure the implementation used satisfies this condition.

Reply:
`NetworkClient` interface only provides the way of `getInputChannel` like existing `addInputChannel` and `removeInputChannel`. Then the decoder would request the buffer from the respective `InputChannel`, so it is not related to the interface whether the buffer is always available. We just migrated the request buffer from `InputChannel` from previous handler to new decode handler.